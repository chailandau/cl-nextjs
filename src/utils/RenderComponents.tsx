import { FC, Fragment } from 'react';

import { Page_PageSections, Project_ProjectSections } from '@/api/graphqlTypes';
import Image from '@/atoms/Image';
import Hero from '@/components/Hero';
import SingleUse from '@/components/SingleUse';
import Socials from '@/components/Socials';
import TextBlock from '@/molecules/TextBlock/TextBlock';

interface Sections {
    /** Sections to render */
    components?: Page_PageSections[] | Project_ProjectSections[];
}

const RenderComponents: FC<Sections> = ({ components }) => {
    if (!components || !components.length) {
        return null;
    }
    const renderComponent = (
        component: Page_PageSections | Project_ProjectSections
    ) => {
        switch (component?.__typename) {
            case 'HeroBlock':
                return <Hero {...component?.hero} />;
            case 'TextBlock':
                return (
                    <TextBlock
                        heading={component?.heading}
                        richText={component?.text}
                    />
                );
            case 'ImageBlock':
                return (
                    <Image
                        src={component?.image?.url || ''}
                        alt={component?.image?.alt || ''}
                        width={component?.image?.width || 1000}
                        height={component?.image?.height || 1000}
                    />
                );
            case 'SocialsBlock':
                return <Socials socials={component?.socials} />;
            case 'SingleUseBlock':
                return <SingleUse component={component?.singleUse} />;
            default:
                return null;
        }
    };

    return (
        <>
            {components?.map((component) => (
                <Fragment key={component?.id}>
                    {renderComponent(component)}
                </Fragment>
            ))}
        </>
    );
};

export default RenderComponents;
