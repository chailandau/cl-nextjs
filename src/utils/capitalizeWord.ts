/**
 * Capitalizes the first letter of a string.
 *
 * @param {string} str - String to be capitalized.
 * @return {string} Capitalized string.
 */
export const capitalizeWord = (str: string): string => {
    if (!str) {
        return str;
    }

    const firstLetter = str.charAt(0).toUpperCase();
    const restOfWord = str.slice(1);

    return firstLetter + restOfWord;
};
