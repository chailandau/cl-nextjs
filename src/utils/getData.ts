import { graphQLClient } from '@/api/graphqlClient';
import { Query } from '@/api/graphqlTypes';

export const getData = async (QUERY: string) =>
    await graphQLClient.request<Query>(QUERY);

export const getPageData = async (QUERY: string, slug: string) =>
    await graphQLClient.request<Query>(QUERY, { slug });
