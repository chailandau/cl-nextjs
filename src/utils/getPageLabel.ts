import { Nav_MenuItems } from '@/api/graphqlTypes';

export const getPageLabel = (menuItem: Nav_MenuItems) => {
    const customLabel =
        menuItem?.overridePageName &&
        menuItem?.customLabel !== null &&
        menuItem.customLabel;

    return menuItem?.page && (customLabel || menuItem?.page?.title);
};
