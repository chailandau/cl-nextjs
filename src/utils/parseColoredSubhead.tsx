import { JSX } from 'react';

import Text, { TextProps } from '@/atoms/Text';

interface ChildSegment {
    text: string;
    bold?: boolean;
}

export interface ColoredSubheadProps {
    children: ChildSegment[];
}

type ColoredSubhead = JSX.Element[] | null | undefined;

/**
 * Parses a colored subhead and returns an array of Text components with appropriate styling.
 *
 * @param coloredSubhead - An array of objects representing the colored subhead.
 * @returns An array of Text components with appropriate styling, or null if the input is falsy.
 */
export const parseColoredSubhead = (
    coloredSubhead: ColoredSubheadProps[]
): ColoredSubhead | null => {
    if (!coloredSubhead) {
        return null;
    }

    const colors = ['aqua', 'white', 'pink'];

    return coloredSubhead.flatMap((segment) => {
        const boldIndexes: number[] = [];

        return segment?.children?.map((child, index) => {
            const isBold = child?.bold;
            let textBgColor;

            if (isBold) {
                boldIndexes.push(index);
                const position = boldIndexes?.indexOf(index);
                textBgColor = colors[position % colors?.length];
            }

            const textProps = {
                as: 'span' as TextProps['as'],
                size: 'md' as TextProps['size'],
                font: 'secondary' as TextProps['font'],
                backgroundColor: textBgColor
            };

            return (
                <Text key={index} {...textProps}>
                    {child?.text}
                </Text>
            );
        });
    });
};
