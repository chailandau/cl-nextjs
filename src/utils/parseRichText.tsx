import { JSX } from 'react';

import styles from '../atoms/Text/Text.module.scss';

import Text from '@/atoms/Text';
import { textSizes } from '@/atoms/Text/Text';

interface ChildSegment {
    text: string;
}

export interface RichTextProps {
    children: ChildSegment[];
}

type RichText = JSX.Element[] | null | undefined;

/**
 * Parses rich text and returns an array of Text components.
 *
 * @param richText - An array of objects representing rich text.
 * @returns An array of Text components, or null if the input is falsy.
 */
export const parseRichText = (
    richText: RichTextProps[],
    size: (typeof textSizes)[number] = 'xs'
): RichText | null => {
    if (!richText) {
        return null;
    }

    const results: JSX.Element[] = [];

    richText.flatMap(
        (segment) =>
            segment?.children?.forEach((child) => {
                const paragraphs = child.text
                    .split('\n')
                    .filter((paragraph) => paragraph.trim() !== '');

                paragraphs.forEach((par, parIndex) => {
                    results.push(
                        <Text
                            className={styles['rich-text']}
                            size={size}
                            key={parIndex}
                        >
                            {par}
                        </Text>
                    );
                });
            })
    );

    return results;
};
