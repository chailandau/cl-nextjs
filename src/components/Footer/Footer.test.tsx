import { composeStory } from '@storybook/react';

import Meta, { Default } from './Footer.stories';

import { testAxeAndSnapshot } from '@/utils/testHelpers';

const Footer = composeStory(Default, Meta);

describe('Footer', () => {
    testAxeAndSnapshot({
        component: <Footer />
    });
});
