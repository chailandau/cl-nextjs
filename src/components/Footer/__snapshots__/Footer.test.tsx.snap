// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer matches snapshot 1`] = `
<DocumentFragment>
  <footer
    class="flex section"
  >
    <div
      class="flex mobile-footer"
    >
      <div
        class="flex nav"
      >
        <ul
          class="menu"
        >
          <li>
            <a
              class="link menu-link"
              href="/"
              rel=""
              target=""
            >
              <span
                class="link-text"
              >
                Work
              </span>
            </a>
          </li>
          <li>
            <a
              class="link menu-link"
              href="/about"
              rel=""
              target=""
            >
              <span
                class="link-text"
              >
                About
              </span>
            </a>
          </li>
          <li>
            <a
              class="link menu-link"
              href="/contact"
              rel=""
              target=""
            >
              <span
                class="link-text"
              >
                Contact
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        class="flex copyright"
      >
        <p
          class="text xs primary"
        >
          copyright chai landau 2023
        </p>
      </div>
    </div>
    <div
      class="flex desktop-footer"
    >
      <div
        class="flex copyright"
      >
        <p
          class="text xs primary"
        >
          copyright chai landau 2023
        </p>
      </div>
      <div
        class="flex nav"
      >
        <ul
          class="menu"
        >
          <li>
            <a
              class="link menu-link"
              href="/"
              rel=""
              target=""
            >
              <span
                class="link-text"
              >
                Work
              </span>
            </a>
          </li>
          <li>
            <a
              class="link menu-link"
              href="/about"
              rel=""
              target=""
            >
              <span
                class="link-text"
              >
                About
              </span>
            </a>
          </li>
          <li>
            <a
              class="link menu-link"
              href="/contact"
              rel=""
              target=""
            >
              <span
                class="link-text"
              >
                Contact
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </footer>
</DocumentFragment>
`;
